src/core/RateSpan.java

```diff
@@ -116,61 +116,81 @@ private void initializeIfNotDone() {
-            + " strictly greater than the previous one (" + t0 + "), but it's"
-            + " not.  this=" + this);
-      }
-      // TODO: for backwards compatibility we'll convert the ms to seconds
-      // but in the future we should add a ratems flag that will calculate
-      // the rate as is.
-      final double time_delta_secs = ((double)(t1 - t0) / 1000.0);
-      double difference;
-      if (prev_data.isInteger() && next_data.isInteger()) {
-        // NOTE: Calculates in the long type to avoid precision loss
-        // while converting long values to double values if both values are long.
-        // NOTE: Ignores the integer overflow.
-        difference = next_data.longValue() - prev_data.longValue();
-      } else {
-        difference = next_data.toDouble() - prev_data.toDouble();
-      }
-      
+    }
+  }
+
+  /**
+   * @return time delta in seconds
+   */
+  private double getTimeDeltaSecs() {
+    // Validate the datapoint first.
+    final long t0 = prev_data.timestamp();
+    final long t1 = next_data.timestamp();
+    if (t1 <= t0) {
+      throw new IllegalStateException(
+          "Next timestamp (" + t1 + ") is supposed to be "
+          + " strictly greater than the previous one (" + t0 + "), but it's"
+          + " not.  this=" + this);
+    }
+    return (double)((t1 - t0) / 1000.0);
+  }
+
+  /**
+   * @return time delta in seconds
+   */
+  private double getValueDifference() {
+    double difference;
+    if (prev_data.isInteger() && next_data.isInteger()) {
+      // NOTE: Calculates in the long type to avoid precision loss
+      // while converting long values to double values if both values are long.
+      // NOTE: Ignores the integer overflow.
+      difference = next_data.longValue() - prev_data.longValue();
+    } else {
+      difference = next_data.toDouble() - prev_data.toDouble();
+    }
+    return difference;
+  }
+
+  /*
+   * Marks the end of rate time series.
+   */
+  private void markEndOfSeries() {
+    // Invalidates the next rate with invalid timestamp.
+    next_rate.reset(INVALID_TIMESTAMP, 0);
+  }
+  /**
+   * Populate the next rate.
+   */
+  private void populateNextRate() {
+    if (source.hasNext()) {
+      moveToNextDatapoint();
+      double difference = getValueDifference();
       if (options.isCounter() && difference < 0) {
-        if (prev_data.isInteger() && next_data.isInteger()) {
-          // NOTE: Calculates in the long type to avoid precision loss
-          // while converting long values to double values if both values are long.
-          difference = options.getCounterMax() - prev_data.longValue() +
-              next_data.longValue();
-        } else {
-          difference = options.getCounterMax() - prev_data.toDouble() +
-              next_data.toDouble();
+        while(source.hasNext()) {
+          moveToNextDatapoint();
+          difference = getValueDifference();
+          if (difference > 0) {
```

Comment:
I think this should be >= 0, currently this just skips until counter increases, which it may not, for some time.